# Final-Year-Project-on-COVID-19
In my final year project for Biomedical Sciences, I evaluated methods to identify SARS-CoV-2 variants with enhanced transmission rates from epidemiological and phylogenetic data. 

Highlight of the project:
- I summarised existing epidemiological methods for calculating relative transmission rates of pathogens from literature, analysed their relevancies to SARS-CoV-2 and the availability of their data inputs.
- I applied a model developed for an oseltamivir-resistant strain of H1N1 influenza that only require publicly available and published data, and obtained the relative transmission rates of SARS-CoV-2 variants such as D614G, Alpha and Beta.
- I modified the model to fit the COVID-19 setting, such as using whole genome sequences instead of genotyping as it provides more insights into the transmission history.
- I compared the inferred values obtained from this model and other published epidemiological models and concluded that this model can generate comparable results using fewer and more accessible data sources.

The repository enclosed part of the codes used for constructing the inference model (e.g. main_all_country.R). Feel free to reach out to me about the methodology and analysis.
